homeassistant/components/keba/__init__.py

```diff
@@ -0,0 +1,304 @@
+
+    await hass.loop.create_datagram_endpoint(lambda: keba,
+                                             local_addr=('0.0.0.0', 7090),
+                                             remote_addr=(host, 7090))
+
+    # Register services to hass
+    async def async_execute_service(call):
+        """Execute a service to KEBA charging station.
+
+        This must be a member function as we need access to the keba
+        object here.
+        """
+        keba = hass.data[DOMAIN]
+        function_name = _SERVICE_MAP[call.service]
+        function_call = getattr(keba, function_name)
+
+        if 'set_max_current' in function_name:
+            if 'current' in call.data:
+                current = float(call.data['current'])
+                if isinstance(current, (int, float)):
+                    hass.async_create_task(function_call(current))
+                else:
+                    _LOGGER.warning("Current is not of type int or float")
+            else:
+                _LOGGER.warning("Current value not given in service call")
+        elif 'set_energy' in function_name:
+            if 'energy' in call.data:
+                energy = call.data['energy']
+                if isinstance(energy, (int, float)):
+                    hass.async_create_task(function_call(energy))
+                else:
+                    _LOGGER.warning("Energy is not of type int or float")
+            else:
+                _LOGGER.warning("Energy value not given in service call")
+        else:
+            hass.async_create_task(function_call())
+
+    # activate failsafe mode if enabled
+    if config[DOMAIN]['failsafe']:
+        failsafe_timeout = config[DOMAIN]['failsafe_timeout']
+        if failsafe_timeout < 10 or failsafe_timeout > 600:
+            _LOGGER.warning("Failsafe timeout not allowed, setting to 30 "
+                            "seconds.")
+            failsafe_timeout = 30
+
+        failsafe_fallback = config[DOMAIN]['failsafe_fallback']
+        if (failsafe_fallback < 6 and failsafe_fallback != 0) \
+           or failsafe_fallback > 63:
+            _LOGGER.warning("Failsafe fallbock current not allowed, setting "
+                            "to 6 A")
+            failsafe_fallback = 6
+
+        failsafe_save = config[DOMAIN]['failsafe_save']
+        if failsafe_save not in [0, 1]:
+            _LOGGER.warning("Failsafe save not allowed, setting to 0")
+            failsafe_save = 0
+
+        keba.send('failsafe ' + str(failsafe_timeout) + ' '
+                  + str(failsafe_fallback * 1000) + ' ' + str(failsafe_save))
+    else:
+        keba.send('failsafe 0 0 0')
+
+    # Register services
+    for service in _SERVICE_MAP:
+        hass.services.async_register(DOMAIN, service, async_execute_service)
+
+    for domain in SUPPORTED_COMPONENTS:
+        hass.async_create_task(
+            discovery.async_load_platform(hass, domain, DOMAIN, {}, config))
+
+    return True
+
+
+class KebaProtocol(asyncio.DatagramProtocol):
```

Comment:
The protocol should be kept in the library. I suggest exposing a client interface from the library.

Reply:
Protcol inlcuding a small warpper for connection creation in asyncio style is now in the library. The code for Home Assistant Integration extends the Wrapper of the library to add soe methods for paramter handling. In this way the async_setup is not so bloated.